### PR TITLE
Automatically focus the WYSIWYG composer when you enter a room

### PR DESCRIPTION
**Pull request by andybalaam**
---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Automatically focus the WYSIWYG composer when you enter a room ([\#9412](https://github.com/matrix-org/matrix-react-sdk/pull/9412)). Contributed by @andybalaam.

### CODE REVIEW

File: src/components/views/rooms/wysiwyg_composer/WysiwygComposer.tsx [hunk 1]
```diff
@@ -51,6 +55,32 @@ export function WysiwygComposer(
         ref.current?.focus();
     }, [content, mxClient, roomContext, wysiwyg, props, ref]);
 
+    useDispatcher(defaultDispatcher, (payload: ActionPayload) => {
```
Comment (florianduros) [inline src/components/views/rooms/wysiwyg_composer/WysiwygComposer.tsx:58]: This hook is a bit massive, if you can make a custom hook from it, it'll be great!

File: src/components/views/rooms/wysiwyg_composer/WysiwygComposer.tsx [hunk 2]
```diff
@@ -51,6 +55,32 @@ export function WysiwygComposer(
...
+                if (context === roomContext.timelineRenderingType) {
+                    // Immediately set the focus, so if you start typing it
+                    // will appear in the composer
+                    ref.current?.focus();
+                    // If we call focus immediate, the focus _is_ in the right
+                    // place, but the cursor is invisible, presumably because
+                    // some other event is still processing.
+                    // The following line ensures that the cursor is actually
+                    // visible in composer.
+                    setTimeout(() => ref.current?.focus(), 200);
```
Comment (florianduros) [inline src/components/views/rooms/wysiwyg_composer/WysiwygComposer.tsx:77]: Maybe we should clear the previous timeout if it's still in waiting state

**Review by florianduros — CHANGES_REQUESTED** [review]

**Review by florianduros — APPROVED** [review]

File: test/components/views/rooms/wysiwyg_composer/WysiwygComposer-test.tsx [hunk 3]
```diff
@@ -144,5 +146,20 @@ describe('WysiwygComposer', () => {
         expect(mockClient.sendMessage).toBeCalledWith('myfakeroom', null, expectedContent);
         expect(screen.getByRole('textbox')).toHaveFocus();
     });
+
+    it('Should focus when receiving an Action.FocusSendMessageComposer action', async () => {
+        // Given we don't have focus
+        customRender(null, false).container;
```
Comment (florianduros) [inline test/components/views/rooms/wysiwyg_composer/WysiwygComposer-test.tsx:152]: I think that the `.container` is not needed here

**Review by florianduros — CHANGES_REQUESTED** [review]

File: test/components/views/rooms/wysiwyg_composer/WysiwygComposer-test.tsx (hunk 3 above)
Comment (andybalaam) [reply test/components/views/rooms/wysiwyg_composer/WysiwygComposer-test.tsx:152]: Ah thanks, that was left over from a previous experiment.

**Review by florianduros — APPROVED** [review]

**Pull request merged**